src/persist/src/file.rs

```diff
@@ -182,6 +183,198 @@ impl Persister for FilePersister {
+        metadata_file.read_to_end(&mut bytes)?;
+        let seqno_start = if bytes.len() != 0 {
+            assert!(bytes.len() == 8);
+            let mut bytes = &bytes[..];
+            let mut buf = [0u8; 8];
+            // Decode sequence number
+            bytes.read_exact(&mut buf)?;
+            SeqNo(u64::from_le_bytes(buf))
+        } else {
+            metadata_file.write_all(&0u64.to_le_bytes())?;
+            metadata_file.sync_all()?;
+            SeqNo(0)
+        };
+
+        // Retrieve the most recently written sequence number in the buffer,
+        // unless the buffer is empty in which case default to 0.
+        let seqno_end = {
+            let len = buffer_file.metadata()?.len();
+
+            if len > 0 {
+                let mut buf = [0u8; 8];
+                buffer_file.seek(SeekFrom::End(8))?;
+                buffer_file.read_exact(&mut buf)?;
+                SeqNo(u64::from_le_bytes(buf))
+            } else {
+                SeqNo(0)
+            }
+        };
+
+        Ok(FileBuffer {
+            dataz: Arc::new(Mutex::new((buffer_file, metadata_file))),
+            seqno: seqno_start..seqno_end,
+            buf: Vec::new(),
+        })
+    }
+
+    fn lockfile_path(base_dir: &Path) -> PathBuf {
+        base_dir.join(Self::LOCKFILE_PATH)
+    }
+
+    fn buffer_path(base_dir: &Path) -> PathBuf {
+        base_dir.join(Self::BUFFER_PATH)
+    }
+
+    fn metadata_path(base_dir: &Path) -> PathBuf {
+        base_dir.join(Self::METADATA_PATH)
+    }
+}
+
+impl Buffer for FileBuffer {
+    fn write_sync(&mut self, buf: Vec<u8>) -> Result<SeqNo, Error> {
+        self.buf.clear();
+        let write_seqno = self.seqno.end;
+        self.seqno = self.seqno.start..SeqNo(self.seqno.end.0 + 1);
+        // Write length prefixed data, and then the sequence number.
+        let len = buf.len() as u64;
+
+        self.buf.extend(&len.to_le_bytes());
+        self.buf.extend(buf);
+        self.buf.extend(&write_seqno.0.to_le_bytes());
+        self.buf.shrink_to_fit();
+
+        let guard = self.dataz.lock()?;
+        let mut buffer_file = &guard.0;
+        buffer_file.write_all(&self.buf)?;
+        buffer_file.sync_all()?;
+        Ok(write_seqno)
+    }
+
+    fn snapshot<F>(&self, mut logic: F) -> Result<Range<SeqNo>, Error>
+    where
+        F: FnMut(SeqNo, &[u8]) -> Result<(), Error>,
+    {
+        let mut bytes = Vec::new();
+
+        {
+            let guard = self.dataz.lock()?;
+            let mut buffer_file = &guard.0;
+            buffer_file.seek(SeekFrom::Start(0))?;
+            buffer_file.read_to_end(&mut bytes)?;
+        }
+
+        let mut bytes = &bytes[..];
+        let mut buf = [0u8; 8];
+        while !bytes.is_empty() {
+            // Decode the data
+            bytes.read_exact(&mut buf)?;
+            let data_len = u64::from_le_bytes(buf);
+            let mut data = vec![0u8; data_len.try_into().unwrap()];
+            bytes.read_exact(&mut data[..])?;
```

Comment:
reuse data to avoid allocating every time? not too worried about this tho, since FileBuffer (and maybe the snapshot method on Buffer) will change at some point

Reply:
I'm going to punt on this for now - will leave a todo comment because in addition to resizing the buffer each time you also need to appropriately fill it with the right number of zeroes (or any other byte pattern) because the read_exact looks to the slice, which holds the pointer to the underlying buffer and the length (not the capacity) to figure out how many bytes to read.

`set_len` also works here but is unsafe and maybe not worth doing for now?

Reply:
Vec has a resize method which I think is what you'd want for this. Definitely not set_len :)

I'm fine leaving this out though.

Reply:
Resize doesn't work here because `read_exact` takes a mutable slice which has to have the exact length, and not the exact capacity for the bytes. I tried that previously when implementing `FileStream` and could not get it to work -- I might still have been missing something though?

Reply:
I don't follow? resize changes the length of the vec not the capacity.

Reply:
Ah my bad! I was mixing up `reserve` and `resize` -- agree that `resize` does what we need to here